Playlists: Manual playlist query performance improvements

bjtitus opened the pull request:
Fixes [PCIOS-221](https://linear.app/a8c/issue/PCIOS-221/improve-performance-with-large-databases)

All changes are behind `optimizeManualPlaylistQueries` feature flag

- `.episode`: Episodes filtered by playlist membership before deduplication window function (adds INNER JOIN playlist to the CTE)
- `.episodeCount` / `.allEpisodeCount`: New inline query implementations using the same optimized deduplication structure
- `manualPlaylistFirstDistinctEpisodes` - Other sorts:
  - Simplified CTE selects only id, podcast_id, playlist_position, rn
  - Joins back to episode table to select episode.* for correct column output

## To test

* Test with a very large database with manual playlists with several thousand items
* ✅ Verify that episode count is the same as before
* ✅ Verify that playlist play time is the same as before
* ✅ Verify that playlist artwork is the same as before

## Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [x] I have considered adding unit tests for my changes.
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

bjtitus committed:
Add feature flag for manual playlist query improvements

# Conflicts:
#	Modules/Utils/Sources/PocketCastsUtils/Feature Flags/FeatureFlag.swift

bjtitus committed:
Optimize manual playlist queries

bjtitus committed:
Add before/after tests for manual playlist query perf

bjtitus committed:
Add changelog entry

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/Automattic/pocket-casts-ios/milestone/111">8.2 ❄️</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

danielebogo commented on Modules/DataModel/Sources/PocketCastsDataModel/Public/PlaylistQueryBuilder.swift:
```diff
@@ -46,6 +47,9 @@ public class PlaylistQueryBuilder {
             switch clause {
             case .episode:
+                // Optimization: Filter episodes before deduplication to reduce window function processing
+                let episodeFilter = FeatureFlag.optimizeManualPlaylistQueries.enabled ?
```
With the FF on, when showing/hiding the archived episodes, the view detail doesn't update correctly:
- Open a manual playlist with few episodes archived
- Show/hide the archived episodes
- The list updates but the artwork doesn't
- Tap it again and notice the artwork updates based on the previous selection
- Sometimes it doesn't even update the artwork
- Switching off or commenting this line makes it work

https://github.com/user-attachments/assets/35a60d14-b91b-40aa-8cec-37376b0baa29

bjtitus replied:
When I tried toggling Show Archived I kept running into what I think was an issue with the artwork loading task in `PlaylistDetailViewModel`. I think instead of blocking with `isLoadingData`, we should just cancel the task instead.

I think there also may have been issues in selecting duplicate episodes with the wrong archived state. I changed the `deduped_episode` portion a bit to prefer episodes with the correct `episode.archived` state.

danielebogo reviewed (changes requested):
@bjtitus there's a problem testing the show/hide archived episodes in manual playlists. This doesn't happen in the current beta or prod.

danielebogo commented:
I added @SergioEstevao as well to make some profiling as well

SergioEstevao commented:
I've run this into the Profiler using the large database on an iPhone 14 Pro running iOS 26.1

Performance is much improved and I don't experience hangs, we still see peaks of CPU usage but it's more distributed and out of the main thread.

<img width="1836" height="1058" alt="Screenshot 2025-12-29 at 14 12 28" src="https://github.com/user-attachments/assets/81ca6b85-e7ef-4ea1-ad63-c5520bea4976" />

bjtitus committed:
Optimize manual playlist queries with improved episode deduplication

- Implement WHERE IN clause for episode deduplication instead of removing filter
- Fix episode deduplication to use firstDistinctEpisodes queries
- Correct archived filter placement in manual playlist queries
- Prefer matching archived status when deduplicating episodes
- Fix artwork updates when toggling archived episodes

bjtitus committed:
Fix off-by-one sync issue with rapid archive toggle

When rapidly toggling the show/hide archived filter, the UI would get "off by one" where the displayed artwork and episode list would be from the previous toggle state, not the current one.

This adds Task cancellation to better handle quick toggles of the show / hide archived button

bjtitus committed:
Use notDownloaded enum value
